packages/rpc-methods/src/permitted/requestSnaps.ts

```diff
@@ -61,30 +63,76 @@ export type RequestSnapsHooks = {
- *
- * Note: Currently, we don't compare caveats, if any caveats are requested, we always return false.
+ * Checks whether an origin has existing `wallet_snap` permission and
+ * whether or not it has the requested snapIds caveat.
  *
  * @param existingPermissions - The existing permissions for the origin.
- * @param requestedPermissions - The requested permissions for the origin.
- * @returns True if the existing permissions satisfy the requested permissions, otherwise false.
+ * @param requestedSnaps - The requested snaps.
+ * @returns True if the existing permissions satisfy the requested snaps, otherwise false.
  */
-function hasPermissions(
+export function hasRequestedSnaps(
   existingPermissions: Record<string, PermissionConstraint>,
-  requestedPermissions: RequestedPermissions,
+  requestedSnaps: Record<string, unknown>,
 ): boolean {
-  return Object.entries(requestedPermissions).every(
-    ([target, requestedPermission]) => {
-      if (
-        requestedPermission?.caveats &&
-        requestedPermission.caveats.length > 0
-      ) {
-        return false;
-      }
-
-      return hasProperty(existingPermissions, target);
+  const snapIdCaveat = existingPermissions[
+    WALLET_SNAP_PERMISSION_KEY
+  ]?.caveats?.find(
+    (caveat: Caveat<string, Json>) => caveat.type === SnapCaveatType.SnapIds,
+  );
+
+  const permittedSnaps = snapIdCaveat?.value;
+  if (isObject(permittedSnaps)) {
+    return Object.keys(requestedSnaps).every((requestedSnap) =>
+      hasProperty(permittedSnaps, requestedSnap),
+    );
+  }
+  return false;
+}
+
+/**
+ * Constructs a valid permission request with merged caveats based on existing permissions
+ * and the requested snaps.
+ *
+ * @param existingPermissions - The existing permissions for the origin.
+ * @param requestedPermissions - The permission request passed into `requestPermissions`.
+ * @returns `requestedPermissions`.
+ */
+export function getSnapPermissionsRequest(
+  existingPermissions: Record<string, PermissionConstraint>,
+  requestedPermissions: Record<string, any>,
+): RequestedPermissions {
+  if (!existingPermissions[WALLET_SNAP_PERMISSION_KEY]) {
+    return requestedPermissions;
+  }
+
+  const snapIdCaveat = existingPermissions[
+    WALLET_SNAP_PERMISSION_KEY
+  ].caveats?.find(
+    (caveat: Caveat<string, Json>) => caveat.type === SnapCaveatType.SnapIds,
+  );
+
+  const permittedSnaps = snapIdCaveat?.value ?? {};
+  const snapIdSet = new Set(Object.keys(permittedSnaps));
+
+  const requestedSnaps =
+    requestedPermissions[WALLET_SNAP_PERMISSION_KEY].caveats[0].value ?? {};
+
+  for (const requestedSnap of Object.keys(requestedSnaps)) {
+    snapIdSet.add(requestedSnap);
+  }
+
+  const mergedCaveatValue = [...snapIdSet].reduce<Record<string, unknown>>(
+    (request, snapId) => {
+      request[snapId] = {};
+      return request;
     },
+    {},
   );
+
+  requestedPermissions[WALLET_SNAP_PERMISSION_KEY].caveats[0].value =
+    mergedCaveatValue;
```

Comment:
Is `caveats[0]` guaranteed to exist here?

Reply:
Can you add some tests for the merging too?

Reply:
yup, `requestedPermissions` is hardcoded before the `getSnapPermissionsRequest` call.

Reply:
There's a test for `getSnapPermissionsRequest`.